src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -0,0 +1,338 @@
+            }
+        }
+
+        public int InstallApk(string apkPath)
+        {
+            _log.LogInformation($"Attempting to install {apkPath}: ");
+            var result = RunAdbCommand($"install {apkPath}");
+            if (result.ExitCode != 0)
+            {
+                _log.LogError($"Error:{Environment.NewLine}{FormatProcessOutputs(result)}");
+            }
+            else
+            {
+                _log.LogInformation($"Successfully installed {apkPath}.");
+            }
+            return result.ExitCode;
+        }
+
+        public int UninstallApk(string apkName)
+        {
+            _log.LogInformation($"Attempting to remove apk '{apkName}': ");
+            var result = RunAdbCommand($"uninstall {apkName}");
+            if (result.ExitCode == 0)
+            {
+                _log.LogInformation($"Successfully uninstalled {apkName}.");
+            }
+            else if (result.ExitCode == 255)
+            {
+                _log.LogInformation($"APK '{apkName}' not on device.");
+            }
+            else
+            {
+                _log.LogError($"Error: {FormatProcessOutputs(result)}");
+            }
+            return result.ExitCode;
+        }
+
+        // This function works but given we'll likely only be using Instrumentations doesn't matter.
+        public int KillApk(string apkName)
+        {
+            _log.LogInformation($"Killing all running processes for '{apkName}': ");
+            var result = RunAdbCommand($"shell am kill --user all {apkName}");
+            if (result.ExitCode != 0)
+            {
+                _log.LogError($"Error:{Environment.NewLine}{FormatProcessOutputs(result)}");
+            }
+            else
+            {
+                _log.LogDebug($"Success!{Environment.NewLine}{result.StandardOutput}");
+            }
+            return result.ExitCode;
+        }
+
+        public void GrantPermissions(string apkName, string[] permissions)
+        {
+            _log.LogInformation($"Granting permissions to '{apkName}' ");
+
+            foreach (string permission in permissions)
+            {
+                var result = RunAdbCommand($"shell pm grant {apkName} {permission}");
+
+                if (result.ExitCode != 0)
+                {
+                    _log.LogError($"Failed granting '{permission}': Exit code: {result.ExitCode}{Environment.NewLine}{result.StandardOutput}; execution may fail as a result.");
+                }
+                else
+                {
+                    _log.LogDebug($"Successfully granted permission:{Environment.NewLine}{permission}");
+                }
+            }
+        }
+
+        // Assumes the directory is empty so any files present after the pull are new.
+        // TODO: pull to %TEMP%\somewhere...\ , count those files and copy over to localPath.
+        public List<string> PullFiles(string devicePath, string localPath)
+        {
+            Directory.CreateDirectory(localPath);

```

Comment:
CreateDirectory can throw a number of exceptions. We might want to look for those and see if there is something we can do.